portfolio/src/main/webapp/mapScript.js

```diff
@@ -0,0 +1,283 @@
+  // Allow users to add markers of their own by clicking on map to show a marker a with text box the user can edit. 
+  map.addListener('click', (event) => {
+    createMarkerForEdit(event.latLng.lat(), event.latLng.lng());
+  });
+
+  // Call function to fetch data of markers to add to map.
+  fetchMarkers();
+
+  // Associate the purple styled map with the MapTypeId and set it to display.
+  map.mapTypes.set('styled_map', purpleStyledMap);
+  map.setMapTypeId('styled_map');
+}
+
+/**
+ * Fetches the data representing locations I have visited from the server and locations other
+ * users have visited from the server to add to the map.
+ */
+function fetchMarkers() {
+  // Fetch locations I have visited and create markers. Pass boolean value true to represent
+  // these are my locations.
+  fetch('/my-markers').then(response => response.json()).then((markers) => {
+    markers.forEach((marker) => {
+      createMarkers(marker.lat, marker.lng, marker.title, marker.content, true)
+    });
+  });
+  // Fetch locations other users have visited and create markers. Pass boolean value false to
+  // represent that these are not my locations.
+  fetch('/user-markers').then(response => response.json()).then((markers) => {
+    markers.forEach((marker) => {
+      createMarkers(marker.lat, marker.lng, marker.title, marker.content, false)
+    });
+  });
+}
+
+/** 
+ * Create a marker from the data fetched to be displayed on map with a read-only info window when clicked.
+ * @param lat: the latitude point of the location
+ * @param lng: the longitude point of the location
+ * @param title: the name of the marker and the title for the info window attached to the marker
+ * @param content: a description of the location for the info window attached to the marker
+ * @param isMyLocation: a boolean that represents if data is for a location I have visited (true) or user has visited (false)
+ */
+function createMarkers(lat, lng, title, content, isMyLocation) {
+
+  // Create a marker with data.
+  const marker = new google.maps.Marker({
+    position: {lat: lat, lng: lng},
+    icon: 'https://maps.google.com/mapfiles/ms/icons/purple-dot.png',
+    map: map,
+    title: title
+  }); 
+  
+  // If marker represents a user inputted location, change icon to pink.
+  if (!isMyLocation) {
+    marker.setIcon('https://maps.google.com/mapfiles/ms/icons/pink-dot.png')
+  }
+
+  // Create a read-only info window (text box that displays title and content description of location).
+  const infoWindow = new google.maps.InfoWindow({
+    content: '<p class="infowindow-title">' + title + '</p>' + content,
+    maxWidth: 400
+  });
+
+  // Add info window to marker that shows when marker is clicked on.
+  marker.addListener('click', () => {
+    infoWindow.open(map, marker);
+  });
+}
+
+/**
+ * Creates a marker that shows a textbox the user can edit.
+ */
+function createMarkerForEdit(lat, lng) {
+  // If we are already showing an editable marker, then remove editable marker.
```

Comment:
When will this happen? ( Like why would the marker already be shown)

Reply:
If I click on somewhere on the map, a temporary editable marker is shown with a textbox window asking me for a title and description of the new marker I am attempting to add. If instead of submitting a marker, I click on another part of the map with a different location than the one I was currently on and was open, this indicates I want to create a marker at this second location so the first editable marker is closed. Clicking the map while an editable marker is open will close the one that was currently open and allow a new editable marker to open at the second location. So if I first clicked on an area in Michigan to create a marker, but change my mind and decide to click on another area by the Colorado border, the Michigan editable marker is closed automatically without me having to click the X at the top right corner and so that I don't have two editable markers open at once. See screenshots:
https://screenshot.googleplex.com/hukj9TmVMNi
https://screenshot.googleplex.com/1t1XhdLxCuC